Code/Tools/AssetProcessor/native/tests/assetmanager/AssetProcessorManagerTest.cpp

```diff
@@ -4226,12 +4226,20 @@ TEST_F(LockedFileTest, DeleteFile_LockedProduct_DeletesWhenReleased)
     if (!AZ::IO::SystemFile::Delete(productPath.toUtf8().constData()))
     {
         AZStd::thread workerThread;
+        bool threadStarted = false;
 
-        m_callback = [&product, &workerThread]() {
-            workerThread = AZStd::thread([&product]() {
-                AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(60));
-                product.close();
-            });
+        m_callback = [&product, &workerThread, &threadStarted]() {
+            // This callback will be called multiple times when AP retries the delete, make sure we only start the thread once
+            if (!threadStarted)
+            {
+                threadStarted = true;
+                workerThread = AZStd::thread(
+                    [&product]()
+                    {
+                        AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(60));
+                        product.close();
+                    });
```

Comment:
Why are we creating a thread to close a file that doesn't get re-open?

When `product.close()` statement completes the file should be closed.
There is no need to make thread.

The callback can just have a single statement of `product.close()`

Reply:
The only reason I can think of is to make it so that the file gets asynchronously closed, ie, to 'prove' that AssessDeletedFile retries and eventually succeeds.
Which seems reasonable to me.  Note:  invokeMethod is async and returns instantly, so no point int putting close after that, and BlockUntilIdle will block, so no point in putting the close after that - you need a thread.

Reply:
actually I'm not sure whats going on here.

I need to look at this again

Reply:
It seems more like what we want is a thread that does this:
* signal that its started
* wait for some amount of time
* close the file
* exit the thread

And we want the main non-threaded body to
* Start the thread
* WAit for the thread to indicate that it has started (threads dont start instantly)
* Tell the APM to do its AssessDeletedFile
* WAit for the APM to finish
* Join the thread
We don't need nested callbacks for this
the body of the callback can just be
```
threadStarted = true;
AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(60));
product.close();
```

And it can be the actual body of the thread function, no need to have a callback that makes a thread itself.

Reply:
Nick pretty much explained it, but the design of this test is that we queue a delete call to the APM (which is not blocking), as part of that, a message is sent saying "hey this file is deleted".  The callback runs in response to that message, which occurs before the delete attempt.  Since we're testing how the AP reacts to a file being locked, we start up a thread so the AP can continue onwards and sleep for some time and then unlock the file.  This way AP has time to attempt at least 1 delete and fail at it before the file is unlocked.

Reply:
It sounds like in that case a one of thread after the `AssessDeletedFile` signal has been sent, that waits 60 milliseconds and then closes the file once should work.

And the thread should join right after the `BlockUntilIdle` call